### PR TITLE
Add basic auth support for Tachidesk

### LINKED ISSUE #9776: Add basic authentication to Tachidesk extension(v1.2.3)
### Describe your suggested feature

Currently if the Tachidesk server has basic authentication set up, there is no way for the extension to access it. 
Only get a HTTP 401 error even with the username and password added to the URL.

It would be great if the Tachidesk extension of Tachiyomi could accept basic auth credentials in it's setting and use them to authenticate.

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I have updated the app to version **[0.12.3](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by JiPaix**
close #9776
close Suwayomi/Tachidesk-Server#247

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt [hunk 1]
```diff
@@ -124,12 +141,27 @@ class Tachidesk : ConfigurableSource, HttpSource() {
...
+    override fun imageRequest(page: Page) = GET(page.imageUrl!!, headers)
+
+    // ------------- Settings -------------
     init {
         val initBaseUrl = Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000).getString(ADDRESS_TITLE, ADDRESS_DEFAULT)!!
-
+        val initBaseLogin = Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000).getString(
+            LOGIN_TITLE, LOGIN_DEFAULT
+        )!!
+        val initBasePassword = Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000).getString(
```
Comment (alessandrojean) [inline src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt:153]: Would be better to make the preferences a class variable at this point.

```kotlin
private val preferences: SharedPreferences by lazy {
    Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
}
```

**Review by alessandrojean — CHANGES_REQUESTED** [review]

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt (hunk 1 above)
Comment (JiPaix) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt]: c961e2c
```kt
    private val preferences: SharedPreferences by lazy {
        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
    }

    init {
        val pairList = ArrayList<Pair<String, String>>()
        pairList.add(Pair(ADDRESS_TITLE, ADDRESS_DEFAULT))
        pairList.add(Pair(LOGIN_TITLE, LOGIN_DEFAULT))
        pairList.add(Pair(PASSWORD_TITLE, PASSWORD_TITLE))

        for (pair in pairList) {
            val initBase = preferences.getString(pair.first, pair.second)!!
            if (initBase.isNotBlank()) {
                refreshCategoryList(initBase)
            }
        }
    }
   ```

Comment (alessandrojean) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt]: Seems good :+1:

Comment (JiPaix) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt]: fixed `pairList.add(Pair(PASSWORD_TITLE, PASSWORD_TITLE))` [38eb4e3](https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11477/commits/38eb4e3964a1c09f90d8942d3e7c7930cb73341d)

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt [hunk 2]
```diff
@@ -124,11 +141,26 @@ class Tachidesk : ConfigurableSource, HttpSource() {
...
+        pairList.add(Pair(LOGIN_TITLE, LOGIN_DEFAULT))
+        pairList.add(Pair(PASSWORD_TITLE, PASSWORD_DEFAULT))
+
+        for (pair in pairList) {
+            val initBase = preferences.getString(pair.first, pair.second)!!
+            if (initBase.isNotBlank()) {
+                refreshCategoryList(initBase)
+            }
         }
     }
```
Comment (alessandrojean) [inline src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt:165]:
```suggestion
    init {
        val preferencesMap = mapOf(
            ADDRESS_TITLE to ADDRESS_DEFAULT,
            LOGIN_TITLE to LOGIN_DEFAULT,
            PASSWORD_TITLE to PASSWORD_DEFAULT
        )
        
        preferencesMap.forEach { key, defaultValue ->
            val initBase = preferences.getString(key, defaultValue)!!
            
            if (initBase.isNotBlank()) {
                refreshCategoryList(initBase)
            }
        }
    }
```

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt [hunk 3]
```diff
@@ -28,20 +30,35 @@ import rx.schedulers.Schedulers
...
+    private fun credentials(): Headers {
+        val headerBuilder: Headers.Builder = Headers.Builder()
+        return if (basePassword.isNotEmpty() && baseLogin.isNotEmpty()) {
+            val credentials = Credentials.basic(baseLogin, basePassword)
+            headerBuilder.add("Authorization", credentials)
+            headerBuilder.build()
+        } else {
+            headers
+        }
+    }
```
[the remark this answers is not in the document]
Comment (JiPaix) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt:57]: works fine

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt (hunk 2 above)
Comment (JiPaix) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt]: :+1: 5031669 68e47e0

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt [hunk 4]
```diff
@@ -28,20 +30,39 @@ import rx.schedulers.Schedulers
...
```
Comment (alessandrojean) [inline src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt:61]: Can now be removed.

**Review by alessandrojean — CHANGES_REQUESTED** [review]

File: src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt (hunk 4 above)
Comment (JiPaix) [reply src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt]: done 10eb179

**Review by alessandrojean — APPROVED** [review]

**Pull request merged**